cmd/process-agent/collector.go

```diff
@@ -258,6 +260,35 @@ func (l *Collector) messagesToResultsQueue(start time.Time, name string, message
 	updateProcContainerCount(messages)
 }
 
+// getRequestID generates a unique identifier (string representation of 64 bits integer) that is composed as follows:
+//	1. 22 bits of the seconds in the current month.
+//	2. 28 bits of hash of the hostname and process agent pid.
+// 	3. 14 bits of the current message in the batch being sent to the server.
+func (l *Collector) getRequestID(start time.Time, chunkIndex int) (string, error) {
+	// The epoch is the beginning of the month of the `start` variable.
+	epoch := time.Date(start.Year(), start.Month(), 1, 0, 0, 0, 0, start.Location())
+	// We are taking the seconds in the current month, and representing them under 22 bits.
+	// In a month we have 60 seconds per minute * 60 minutes per hour * 24 hours per day * maximum 31 days a month
+	// which is 2678400, and it can be represented with log2(2678400) = 21.35 bits.
+	seconds := fmt.Sprintf("%022b", int64(start.Sub(epoch).Seconds()))
+	// Next, we want 28 bits of hashed hostname & process agent pid.
+	hash := fnv.New32()
+	hash.Write([]byte(l.cfg.HostName))
+	hash.Write([]byte(strconv.Itoa(os.Getpid())))
+	hostNamePIDHash := fmt.Sprintf("%032b", hash.Sum32())
+
+	// Next, we take up to 14 bits to represent the message index in the batch.
+	// It means that we support up to 16384 (2 ^ 14) different messages being sent on the same batch.
+	chunk := fmt.Sprintf("%014b", chunkIndex)
+	requestID := seconds[0:22] + hostNamePIDHash[0:28] + chunk[0:14]
```

Comment:
this seems like a weird way to write this since `seconds` and `chunk` are guaranteed to be 22 and 14 bytes respectively

Reply:
`%014b` guarantees to have at least 14 characters, if we need 10, it will pad the result with 4 leading zeros, but if we need more than 14 bits, it will not pad and return the result as is
https://go.dev/play/p/HFiqoRUn1qJ